### ISSUE TITLE
NullPointerException: Deflater has been closed only with GzipResponseFilter

### ISSUE THREAD

**Issue by matinh** [omnifaces/omnifaces#768]
I'm seeing occasional exceptions like this in my log-file:

```
2023-08-09 14:36:49,096 ERROR [io.undertow.request] (default task-1) UT005023: Exception handling request to /XXX/javax.faces.resource/touch/touchswipe.js.xhtml: javax.servlet.ServletException: java.lang.NullPointerException: Deflater has been closed
        at deployment.XXX.ear//org.apache.shiro.web.servlet.AdviceFilter.cleanup(AdviceFilter.java:196)
        at deployment.XXX.ear//org.apache.shiro.web.servlet.AdviceFilter.doFilterInternal(AdviceFilter.java:148)
        at deployment.XXX.ear//org.apache.shiro.web.servlet.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:154)
        [...]
Caused by: java.lang.NullPointerException: Deflater has been closed
        at java.base/java.util.zip.Deflater.ensureOpen(Deflater.java:898)
        at java.base/java.util.zip.Deflater.deflate(Deflater.java:566)
        at java.base/java.util.zip.Deflater.deflate(Deflater.java:465)
        at java.base/java.util.zip.GZIPOutputStream.finish(GZIPOutputStream.java:163)
        at java.base/java.util.zip.DeflaterOutputStream.close(DeflaterOutputStream.java:244)
        at deployment.XXX.war//org.omnifaces.io.ResettableBufferedOutputStream.close(ResettableBufferedOutputStream.java:162)
        at deployment.XXX.war//org.omnifaces.io.ResettableBufferedOutputStream.close(ResettableBufferedOutputStream.java:162)
        at deployment.XXX.war//org.omnifaces.io.DefaultServletOutputStream.close(DefaultServletOutputStream.java:72)
        at deployment.XXX.war//org.omnifaces.servlet.HttpServletResponseOutputWrapper.close(HttpServletResponseOutputWrapper.java:140)
        at deployment.XXX.war//org.omnifaces.servlet.GzipHttpServletResponse.close(GzipHttpServletResponse.java:141)
        at deployment.XXX.war//org.omnifaces.filter.GzipResponseFilter.doFilter(GzipResponseFilter.java:184)
        at deployment.XXX.war//org.omnifaces.filter.HttpFilter.doFilter(HttpFilter.java:108)
```

I currently can't reproduce the exception, but it definitely **only** occurs with `GzipResponseFilter` enabled.
And it only occurs for resources like CSS or JavaScript files which go through the JSF ResourceHandler.

I found a similar issue for the Jetty webserver: https://github.com/eclipse/jetty.project/issues/3892. But I don't use Jetty. I'm on Wildfly 26.1.3 with Undertow. With a JSF 2.3 WebApp (Mojarra 2.3.17.SP01). Currently I'm still on OmniFaces version 3.14.2 and I'm using PrimeFaces 8.0 in case it matters.

I enabled the GzipResponseFilter in `web.xml` as follows:

```
  <filter>
    <filter-name>gzipResponseFilter</filter-name>
    <filter-class>org.omnifaces.filter.GzipResponseFilter</filter-class>
  </filter>
  <filter-mapping>
    <filter-name>gzipResponseFilter</filter-name>
    <url-pattern>*.xhtml</url-pattern>
    <dispatcher>REQUEST</dispatcher>
  </filter-mapping>
```

After having a look at the code, it seems closing an already closed stream is the actual problem. My impression is, this happens, when a browser does an up-to-date check for resources (stylesheets or JavaScript files) and aborts the download of the actual resource. This also somehow corresponds with the mentioned Jetty-bug.

Could an additional check for the stream is still open before closing it, or a try-catch around the `close()` in `GzipResponseFiler` be a solution here`?

Thanks for your help and let me know, if you need any additional information.

**Comment by melloware** [comment]
Great debugging. I have seen this error occasionally and could never figure it out.

**Comment by BalusC** [comment]
This seems to be JVM dependent as well. I guess best what I can do is to add a WARN log line with elaborate diagnostic so a future occurrence can be better analyzed.

**Comment by matinh** [comment]
If it is of any help, I'm currently on OpenJDK 17:

```
openjdk version "17.0.7" 2023-04-18
OpenJDK Runtime Environment (build 17.0.7+7-Ubuntu-0ubuntu122.04.2)
OpenJDK 64-Bit Server VM (build 17.0.7+7-Ubuntu-0ubuntu122.04.2, mixed mode, sharing)
```

**Comment by BalusC** [comment]
I've been able to reproduce it (WildFly 26.1.3 + OpenJDK 17.0.2).

1. Put breakpoint on `out.close()` call in `ResourceHandlerImpl` (Mojarra)
2. Open a minimal page with just a single h:commandButton+f:ajax (will load jsf.js)
3. When breakpoint is hit on that jsf.js request, press Esc in browser to cancel request (will basically trigger 'connection reset by peer' condition)
4. Release breakpoint, exact this exception will be thrown

Upon investigation it appears that JDK-specific internal impl of `java.util.zip.DeflaterOutputStream` is since about OpenJDK 11.0.something changed that its own 'closed' flag is only called when it didn't throw exception. And probably other JDKs have their own impl/idea on this. So calling `DeflaterOutputStream#close()` for the second time would throw `java.lang.NullPointerException: Deflater has been closed`. And this was further backtracked to exactly the same issue in OmniFaces own `ResettableBufferedOutputStream`. Basically its `closed` flag is only set when the `close()` has completed successfully while on a second thought it should actually have been set as soon as the `closed()` method is invoked. So it has been fixed that way. The `DeflaterOutputStream` should in JDK should probably be fixed to follow same ideology but that's beyond the responsiblity of OmniFaces ;)
